tests/Unit/AnonymousQuestionTest.php

```diff
@@ -0,0 +1,337 @@
+            Carbon::now()->addMinute(11)
+        );
+    }
+
+    /**
+     * Tests answering a question belonging to the current semester
+     * but before the start date of the evaluation form
+     * (which should fail).
+     * @return void
+     */
+    public function test_answering_not_yet_opened_question(): void
+    {
+        $user = User::factory()->hasEducationalInformation()->create();
+
+        $semester = Semester::current();
+        $question = Question::factory()
+            ->for($semester, 'parent')
+            ->hasOptions(3)
+            ->create();
+
+        $this->delayForm(); // important!
+        // don't forget this
+        $question->refresh();
+
+        $this->expectException(\Exception::class);
+        $question->storeAnswers(
+            $user,
+            $question->options->first(),
+            AnswerSheet::createForUser($user, $semester)
+        );
+    }
+
+    /**
+     * Tests answering the same question twice as the same user
+     * (which should fail).
+     * @return void
+     */
+    public function test_answering_twice(): void
+    {
+        $user = User::factory()->hasEducationalInformation()->create();
+
+        $semester = Semester::current(); // gets created if does not already exist
+        $question = Question::factory()
+            ->for($semester, 'parent')
+            ->hasOptions(3)
+            ->create();
+
+        $this->openForm(); // this also sets the start and end dates
+        $question->refresh();
+
+        $this->expectException(\Exception::class);
+
+        $answerSheet1 = AnswerSheet::createForUser($user, $semester);
+        $answerSheet2 = AnswerSheet::createForUser($user, $semester);
+        $question->storeAnswers($user, $question->options->random(2)->all(), $answerSheet1);
+        $question->storeAnswers($user, $question->options->random(), $answerSheet2);
+    }
+
+    /**
+     * Tests answering a single-choice question.
+     * @return void
+     */
+    public function test_answering_radio(): void
+    {
+        $user = User::factory()->hasEducationalInformation()->create();
+
+        $semester = Semester::current();
+        $question = Question::factory()
+            ->for($semester, 'parent')
+            ->hasOptions(3)
+            ->create(['max_options' => 1]);
+
+        $this->openForm();
+        $question->refresh();
+
+        $answerSheet = AnswerSheet::createForUser($user, $semester);
+        $question->storeAnswers($user, $question->options->first(), $answerSheet);
+
+        $question->load('users');
+
+        echo $question->users()->count() . "\n";
+
+        $this->assertEquals(1, $question->options->first()->votes);
+        $this->assertEquals(0, $question->options->get(1)->votes);
+        $this->assertEquals(0, $question->options->get(2)->votes);
+
+        $this->assertTrue(
+            $question->users()
+                ->where('id', $user->id)
+                ->exists()
+        );
```

Comment:
It's here. The assertTrue sometimes fails and sometimes not; if it fails, it believes there are no users belonging to the question (it echoes 0). I've tried `$question->refresh()` and `$question->load('users')`, but it remains the same. The most confusing thing is its randomness; I've had a similar problem with the opening dates, but there, `$question->refresh()` solved it.

Reply:
Now, I've tried
```php
DB::table('question_user')->insert(['question_id' => $this->id, 'user_id' => $user->id]);
echo DB::table('question_user')->where('question_id', $this->id)->where('user_id', $user->id)->count() . "\n";
```
and it sometimes echoes 0. I'm clueless:D